File: mascota/src/Infraestructure/message_broker/RabbitMQConsumer.ts
import { Channel } from "amqplib";
import { Pet } from "../../Domain/entities/Pet";
import { IPetRepository } from "../../Domain/Repositories/IPetRepository";

export class RabbitMQPetConsumer {
  constructor(private readonly channel: Channel, private readonly petRepository: IPetRepository) {}

  async startConsuming(): Promise<void> {
    const queue = "service_pet";

    // Asegurar que la cola existe
    await this.channel.assertQueue(queue, { durable: true });

    // Escuchar mensajes en la cola
    this.channel.consume(queue, async (message) => {
        if (message) {
          try {
            const eventPayload = JSON.parse(message.content.toString());
            console.log("Payload recibido:", eventPayload);
      
            // Validar que los campos no estén vacíos o undefined
            if (!eventPayload.name || !eventPayload.species || !eventPayload.userId) {
              throw new Error("Faltan campos obligatorios en el evento.");
            }
      
            // Crear la nueva mascota
            const pet = new Pet(
              undefined, // ID generado automáticamente en la base de datos
              eventPayload.name,
              eventPayload.species,
              eventPayload.breed,
              eventPayload.birth_date,
              eventPayload.weight,
              eventPayload.height,
              eventPayload.gender,
              eventPayload.vaccines,
              eventPayload.allergies,
              eventPayload.sterilized,
              eventPayload.userId,
              eventPayload.imageUrl // Guardar la URL de la imagen

            );
      
            // Guardar en la base de datos
            await this.petRepository.create(pet);
      
            // Confirmar el mensaje como procesado
            this.channel.ack(message);
          } catch (error) {
            console.error("Error al procesar el mensaje de RabbitMQ:", error);
            this.channel.nack(message, false, true); // Re-encolar el mensaje
          }
        }
      });
      
  }
}

File: mascotapy/src/Mascota/Infraestructure/message_broker/RabbitMQConsumer.py
import json
import pika
from Mascota.Domain.Entities.Pet import Pet
from Mascota.Domain.Repositories.IPetRepository import IPetRepository


class RabbitMQPetConsumer:
    def __init__(self, channel: pika.adapters.blocking_connection.BlockingChannel, pet_repository: IPetRepository):
        """
        Inicializa el consumidor de RabbitMQ.
        :param channel: Canal de RabbitMQ ya conectado.
        :param pet_repository: Repositorio para manejar las operaciones con mascotas.
        """
        self.channel = channel
        self.pet_repository = pet_repository

    def start_consuming(self):
        """
        Inicia el consumo de mensajes desde la cola 'service_pet'.
        """
        queue = "service_pet"

        try:
            # Declarar la cola en RabbitMQ (asegurar que existe)
            self.channel.queue_declare(queue=queue, durable=True)
            print(f"✔️ Cola '{queue}' declarada o verificada con éxito.")

            # Definir el callback para procesar mensajes
            def callback(ch, method, properties, body):
                try:
                    # Deserializar el mensaje JSON recibido
                    event_payload = json.loads(body)
                    print("📥 Payload recibido:", event_payload)

                    # Validar que los campos obligatorios estén presentes
                    if not event_payload.get("id") or not event_payload.get("name") or not event_payload.get("species") or not event_payload.get("userId"):
                        raise ValueError("❌ Faltan campos obligatorios en el evento.")

                    # Crear la nueva mascota con todos los campos necesarios
                    pet = Pet(
                        id=event_payload.get("id"),  # Pasar el 'id' desde el payload
                        name=event_payload.get("name"),
                        species=event_payload.get("species"),
                        breed=event_payload.get("breed"),
                        birth_date=event_payload.get("birth_date"),
                        weight=float(event_payload.get("weight", 0.0)),  # Convertir weight a float, con valor por defecto 0.0 si no existe
                        height=float(event_payload.get("height", 0.0)),  # Convertir height a float, con valor por defecto 0.0 si no existe
                        gender=event_payload.get("gender"),
                        allergies=event_payload.get("allergies"),
                        sterilized=event_payload.get("sterilized") == 'true',  # Convertir a booleano
                        user_id=event_payload.get("userId"),  # Asegúrate de que el campo userId esté presente
                        image_url=event_payload.get("imageUrl")  # Image URL si está presente
                    )

                    # Llamar al repositorio para almacenar la mascota
                    self.pet_repository.create(pet)
                    print(f"✔️ Mascota '{pet.name}' guardada exitosamente.")

                except ValueError as e:
                    print(f"❌ Error al procesar el mensaje: {str(e)}")
                except Exception as e:
                    print(f"❌ Error desconocido al procesar el mensaje: {str(e)}")

                # Confirmar que el mensaje ha sido procesado
                ch.basic_ack(delivery_tag=method.delivery_tag)

            # Iniciar el consumo de mensajes
            self.channel.basic_consume(queue=queue, on_message_callback=callback)
            print("✔️ Esperando mensajes en la cola. Para salir presiona CTRL+C.")
            self.channel.start_consuming()

        except Exception as e:
            print(f"❌ Error al iniciar el consumidor: {str(e)}")
